corehq/apps/case_importer/tracking/models.py

```diff
@@ -23,6 +27,29 @@ class Meta(object):
     def task(self):
         return get_task(self.task_id)
 
+    @memoized
+    def get_task_status_json(self):
+        if self.task_status_json:
+            return TaskStatus.wrap(self.task_status_json)
+        else:
+            return get_task_status_json(str(self.task_id))
+
+    def set_task_status_json_if_finished(self):
+        """
+        set task_status_json based on self.task_id
+
+        :return: True if the caller must save model to persist the effect
+
+        """
+        if self.task_status_json is None:
+            # intentionally routing through method to prime local cache
+            task_status_json = self.get_task_status_json()
+            if task_status_json.is_finished():
+                self.task_status_json = task_status_json
+                return True
```

Comment:
I'm not sure what the best practice is, but I try to not have models save themselves since that steals the ability of the caller to optimize saves. Even if in this case there's no optimization done, I still like being able to tell from the calling code when something was actually saved. When I'm reading code that interacts with models this is constantly going through my head:

```
my_model.do_something()  # is it saved yet??
my_model.save()  # oh, whew, ok here's where it's saved
```
so I like to see that save there. Might just be personal preference; I don't recall ever having formalized this as a best practice even in conversation.